blocks/library/paragraph/index.js

```diff
@@ -152,39 +153,41 @@ registerBlockType( 'core/paragraph', {
 					</PanelBody>
 				</InspectorControls>
 			),
-			<BlockAutocomplete key="editable" onReplace={ onReplace }>
-				<Editable
-					tagName="p"
-					className={ classnames( 'wp-block-paragraph', className, {
-						[ `align${ width }` ]: width,
-						'has-background': backgroundColor,
-					} ) }
-					style={ {
-						backgroundColor: backgroundColor,
-						color: textColor,
-						fontSize: fontSize ? fontSize + 'px' : undefined,
-						textAlign: align,
-					} }
-					value={ content }
-					onChange={ ( nextContent ) => {
-						setAttributes( {
-							content: nextContent,
-						} );
-					} }
-					focus={ focus }
-					onFocus={ setFocus }
-					onSplit={ ( before, after, ...blocks ) => {
-						setAttributes( { content: before } );
-						insertBlocksAfter( [
-							...blocks,
-							createBlock( 'core/paragraph', { content: after } ),
-						] );
-					} }
-					onMerge={ mergeBlocks }
-					onReplace={ onReplace }
-					placeholder={ placeholder || __( 'New Paragraph' ) }
+			<UserAutocomplete key="editable">
+				<BlockAutocomplete onReplace={ onReplace }>
```

Comment:
I haven't looked in details yet, but I'm wondering if we should create a separate Component for this or enhance the BlockAutocomplete component instead. I guess will have `tags` autocomplete too, so we may end-up with several levels of autocomplete.

Trying to think about what the API should look like, maybe I see something like this:

```js
const blockPattern = {
  triggerOn: ( content ) => content.length === '', // when to trigger the autocomplete could use regex for instance
  getSuggestions: ( content ) => fetch( '/users' ), // A function triggers on each content change if the the triggerOn is true, it returns the list of suggestions (or a promise of list of suggestions)
  renderSuggestion: ( suggestion ) => <Button>{ suggestion.tittle }</Button> // Renders a unique suggestion
  onSelect: ( suggestion ) => // What happens when you select a suggestion
}

const patterns =[ blockPattern, userPattern, tagsPattern ];

<Autocomplete patterns={ patterns } />
```

And the patterns could be declared in separate files for reusability.

Thoughts @aduth

Reply:
The API you suggest is insufficient. If you want it to work for triggers like '@' anywhere in the content then you also need to know the location of the cursor otherwise you will popup dialogs for items that match the trigger but you are nowhere near.

Reply:
Also for things like the `@user` popups you only want them to happen outside of links because a link inside a link is invalid HTML so you have to work with the actual HTML not just the text.

Reply:
Anyway, I am working on making it work like you suggest though I won't have it finished today.

Reply:
Yes, I'm aware the API is probably not feature-proof but just wanted to show an idea on how we can build this generically. (which would ease extending these for plugins as well).

Reply:
I have updated the design of Autocomplete to take in multiple "completers".